matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/workers/SessionKeepAliveScheduler.java

```diff
@@ -1,39 +1,86 @@
+import com.matchbook.sdk.core.exceptions.MatchbookSDKHttpException;
+import com.matchbook.sdk.rest.ConnectionManager;
 import com.matchbook.sdk.rest.UserClient;
+import com.matchbook.sdk.rest.UserClientRest;
 import com.matchbook.sdk.rest.dtos.user.Login;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class SessionKeepAliveScheduler implements Runnable {
+public class SessionKeepAliveScheduler {
+
+    private static final long KEEP_ALIVE_PERIOD_HOURS = 4L;
 
-    private static final Logger LOG = LoggerFactory.getLogger(SessionKeepAliveScheduler.class);
+    private static Logger LOG = LoggerFactory.getLogger(SessionKeepAliveScheduler.class);
 
+    private final long loginTimeout;
+    private final ScheduledExecutorService sessionKeepAliveExecutor;
     private final UserClient userClient;
 
-    public SessionKeepAliveScheduler(UserClient userClient) {
-        this.userClient = userClient;
+    public SessionKeepAliveScheduler(ConnectionManager connectionManager) {
+        loginTimeout = connectionManager.getClientConfig().getHttpConfig().getReadTimeout();
+        this.userClient = new UserClientRest(connectionManager);
+
+        this.sessionKeepAliveExecutor = Executors.newSingleThreadScheduledExecutor(runnable -> {
+            Thread thread = new Thread(runnable);
+            thread.setDaemon(true);
+            thread.setName("mb-sdk-session-manager");
+            return thread;
+        });
+    }
+
+    public void start() {
+        doLogin();
+        startScheduler();
+    }
+
+    private void startScheduler() {
+        SessionKeepAliveTask sessionKeepAliveTask = new SessionKeepAliveTask(userClient);
+        sessionKeepAliveExecutor.scheduleAtFixedRate(sessionKeepAliveTask,
+                KEEP_ALIVE_PERIOD_HOURS,
+                KEEP_ALIVE_PERIOD_HOURS,
+                TimeUnit.HOURS);
+    }
+
+    public void stop() {
+        sessionKeepAliveExecutor.shutdown();
     }
 
-    @Override
-    public void run() {
+    private void doLogin() {
+        final CountDownLatch countDownLatch = new CountDownLatch(1);
+
         userClient.login(new StreamObserver<Login>() {
+
             @Override
             public void onNext(Login login) {
-                //do nothing
+                LOG.info("User {} successfully logged in.", login.getAccount().getUsername());
             }
 
             @Override
             public void onCompleted() {
-                LOG.info("Session token successfully updated");
+                countDownLatch.countDown();
             }
 
             @Override
             public <E extends MatchbookSDKException> void onError(E exception) {
-                LOG.warn("Fail to update session token", exception);
+                throw new MatchbookSDKHttpException(exception, ErrorType.UNAUTHENTICATED);
             }
         });
+
+        try {
+            countDownLatch.await(loginTimeout, TimeUnit.MILLISECONDS);
```

Comment:
i don't really like this one, because if you will change the timeunit on httpconfig it will be not obvious that you should do that in this class too.

Reply:
In truth, I have thought about linking the two already. I will do it since also you noticed that.

Reply:
I change my comment here. I thought about something different in the previous one :slightly_smiling_face: Why you think the two timeouts shouldn't be related? And currently when you modify the read timeout in the `HttpConfig` you don't have to change it in this class too. Quite the opposite, in fact, you'd change it in the keep-alive task too automatically.

Reply:
you right, the number should be the same, but you might change the time units on a code refactor. So, we have two options: sync all time unit + number or remote that dependency

Reply:
Don't we already know our `HttpConfig` uses only milliseconds as time unit?

Anyway, I'm OK to remove the dependency if you are not sure about that. It's not a big deal and honestly I think it's just a minor detail. I just want to make sure our understanding is the same and I'm not missing anything obvious here.